File: app/src/main/java/com/example/myanimedata/activity/SearchActivity.java
package com.example.myanimedata.activity;

import android.os.Bundle;
import android.view.View;
import android.widget.ProgressBar;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.core.text.HtmlCompat;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.example.myanimedata.R;
import com.example.myanimedata.adapter.AnimeAdapter;
import com.example.myanimedata.adapter.CharacterAdapter;
import com.example.myanimedata.adapter.MangaAdapter;
import com.example.myanimedata.api.AnimeResponse;
import com.example.myanimedata.api.AnimeResult;
import com.example.myanimedata.api.ApiClient;
import com.example.myanimedata.api.ApiService;
import com.example.myanimedata.api.CharacterResponse;
import com.example.myanimedata.api.CharacterResult;
import com.example.myanimedata.api.MangaResponse;
import com.example.myanimedata.api.MangaResult;

import java.util.ArrayList;
import java.util.List;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import retrofit2.Retrofit;

public class SearchActivity extends AppCompatActivity {
    private ApiService apiService;
    private AnimeAdapter searchAnimeAdapter;
    private MangaAdapter searchMangaAdapter;
    private CharacterAdapter searchCharacterAdapter;

    private final List<AnimeResult> searchAnimeResults = new ArrayList<>();
    private final List<MangaResult> searchMangaResults = new ArrayList<>();
    private final List<CharacterResult> searchCharacterResults = new ArrayList<>();

    private TextView noSearchResult;
    private RecyclerView rvSearch;

    private ProgressBar progressSearch;

    private int page = 1;
    private String query;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_search);

        query = getIntent().getStringExtra("q");
        String type = getIntent().getStringExtra("type");

        Toolbar searchToolbar = findViewById(R.id.searchToolbar);
        TextView searchResult = findViewById(R.id.textSearchResult);
        noSearchResult = findViewById(R.id.textNoSearchResult);
        rvSearch = findViewById(R.id.rvSearchList);
        progressSearch = findViewById(R.id.loadingSearch);

        setSupportActionBar(searchToolbar);
        setSearchText(searchResult, query);

        Retrofit retrofit = ApiClient.getClient();
        apiService = retrofit.create(ApiService.class);

        GridLayoutManager mLayoutManager = new GridLayoutManager(this, 3);

        rvSearch.setLayoutManager(mLayoutManager);

        switch (type) {
            case "anime":
                searchAnimeAdapter = new AnimeAdapter(searchAnimeResults, this);
                rvSearch.setAdapter(searchAnimeAdapter);
                searchAnimeData(page);

                rvSearch.addOnScrollListener(new RecyclerView.OnScrollListener(){
                    @Override
                    public void onScrolled(@NonNull RecyclerView recyclerView, int dx, int dy){
                        super.onScrolled(recyclerView, dx, dy);

                        if(!recyclerView.canScrollVertically(1)){
                            page++;
                            searchAnimeData(page);
                        }
                    }
                });
                break;

            case "manga":
                searchMangaAdapter = new MangaAdapter(searchMangaResults, this);
                rvSearch.setAdapter(searchMangaAdapter);
                searchMangaData(page);

                rvSearch.addOnScrollListener(new RecyclerView.OnScrollListener(){
                    @Override
                    public void onScrolled(@NonNull RecyclerView recyclerView, int dx, int dy){
                        super.onScrolled(recyclerView, dx, dy);

                        if(!recyclerView.canScrollVertically(1)){
                            page++;
                            searchMangaData(page);
                        }
                    }
                });
                break;

            case "character":
                searchCharacterAdapter = new CharacterAdapter(searchCharacterResults, this);
                rvSearch.setAdapter(searchCharacterAdapter);
                searchCharacterData(page);

                rvSearch.addOnScrollListener(new RecyclerView.OnScrollListener(){
                    @Override
                    public void onScrolled(@NonNull RecyclerView recyclerView, int dx, int dy){
                        super.onScrolled(recyclerView, dx, dy);

                        if(!recyclerView.canScrollVertically(1)){
                            page++;
                            searchCharacterData(page);
                        }
                    }
                });
                break;
        }

        searchToolbar.setOnClickListener(v->onBackPressed());
    }

    private void searchAnimeData(int page) {
        int limit = 15;

        Call<AnimeResponse> call = apiService.searchAllAnime(page, limit, query);
        call.enqueue(new Callback<AnimeResponse>() {
            @Override
            public void onResponse(@NonNull Call<AnimeResponse> call, @NonNull Response<AnimeResponse> response) {
                if(response.body() != null){
                    int oldCount = searchAnimeResults.size();
                    progressSearch.setVisibility(View.GONE);
                    rvSearch.setVisibility(View.VISIBLE);

                    searchAnimeResults.addAll(response.body().getAnimeResults());
                    searchAnimeAdapter.notifyItemRangeInserted(oldCount, searchAnimeResults.size());
                } else if(searchAnimeResults.isEmpty()) {
                    progressSearch.setVisibility(View.GONE);
                    noSearchResult.setVisibility(View.VISIBLE);
                } else {
                    progressSearch.setVisibility(View.GONE);
                }
            }

            @Override
            public void onFailure(@NonNull Call<AnimeResponse> call, @NonNull Throwable t) {
                progressSearch.setVisibility(View.GONE);
                Toast.makeText(SearchActivity.this, "Fail to Fetch Data !!!",
                        Toast.LENGTH_SHORT).show();
            }
        });
    }

    private void searchMangaData(int page) {
        int limit = 15;

        Call<MangaResponse> call = apiService.searchAllManga(page, limit, query);
        call.enqueue(new Callback<MangaResponse>() {
            @Override
            public void onResponse(@NonNull Call<MangaResponse> call, @NonNull Response<MangaResponse> response) {
                if(response.body() != null){
                    int oldCount = searchMangaResults.size();
                    progressSearch.setVisibility(View.GONE);
                    rvSearch.setVisibility(View.VISIBLE);

                    searchMangaResults.addAll(response.body().getMangaResults());
                    searchMangaAdapter.notifyItemRangeInserted(oldCount, searchMangaResults.size());
                } else if(searchMangaResults.isEmpty()){
                    progressSearch.setVisibility(View.GONE);
                    noSearchResult.setVisibility(View.VISIBLE);
                } else {
                    progressSearch.setVisibility(View.GONE);
                }
            }

            @Override
            public void onFailure(@NonNull Call<MangaResponse> call, @NonNull Throwable t) {
                progressSearch.setVisibility(View.GONE);
                Toast.makeText(SearchActivity.this, t.getMessage() + " cause : "
                        + t.getCause(), Toast.LENGTH_SHORT).show();
            }
        });
    }

    private void searchCharacterData(int page) {
        int limit = 15;

        Call<CharacterResponse> call = apiService.searchAllCharacter(page, limit, query);
        call.enqueue(new Callback<CharacterResponse>() {
            @Override
            public void onResponse(@NonNull Call<CharacterResponse> call, @NonNull Response<CharacterResponse> response) {
                if(response.body() != null){
                    int oldCount = searchCharacterResults.size();
                    progressSearch.setVisibility(View.GONE);
                    rvSearch.setVisibility(View.VISIBLE);

                    searchCharacterResults.addAll(response.body().getCharacterResults());
                    searchCharacterAdapter.notifyItemRangeInserted(oldCount, searchCharacterResults.size());
                } else if(searchCharacterResults.isEmpty()) {
                    progressSearch.setVisibility(View.GONE);
                    noSearchResult.setVisibility(View.VISIBLE);
                } else {
                    progressSearch.setVisibility(View.GONE);
                }
            }

            @Override
            public void onFailure(@NonNull Call<CharacterResponse> call, @NonNull Throwable t) {
                progressSearch.setVisibility(View.GONE);
                Toast.makeText(SearchActivity.this, "Fail to Fetch Data !!!",
                        Toast.LENGTH_SHORT).show();
            }
        });
    }

    private void setSearchText(TextView tv, String textValue){
        tv.setText(HtmlCompat.fromHtml("Search Result For : <b>' " + textValue + " '</b>", HtmlCompat.FROM_HTML_MODE_LEGACY));
    }
}

File: app/src/main/java/com/example/myanimedata/activity/SearchActivity.kt
package com.example.myanimedata.activity

import android.content.res.Configuration
import android.os.Bundle
import android.view.View
import android.widget.ProgressBar
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import androidx.core.text.HtmlCompat
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.example.myanimedata.R
import com.example.myanimedata.adapter.AnimeAdapter
import com.example.myanimedata.adapter.CharacterAdapter
import com.example.myanimedata.adapter.MangaAdapter
import com.example.myanimedata.api.*
import com.example.myanimedata.api.ApiClient.client
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class SearchActivity : AppCompatActivity() {
    private var apiService: ApiService? = null
    private var searchAnimeAdapter: AnimeAdapter? = null
    private var searchMangaAdapter: MangaAdapter? = null
    private var searchCharacterAdapter: CharacterAdapter? = null
    private val searchAnimeResults: MutableList<AnimeResult> = ArrayList()
    private val searchMangaResults: MutableList<MangaResult> = ArrayList()
    private val searchCharacterResults: MutableList<CharacterResult> = ArrayList()
    private lateinit var noSearchResult: TextView
    private lateinit var rvSearch: RecyclerView
    private lateinit var progressSearch: ProgressBar
    private var page = 1
    private var query: String? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_search)
        query = intent.getStringExtra("q")

        val type = intent.getStringExtra("type")
        val searchToolbar = findViewById<Toolbar>(R.id.searchToolbar)
        val searchResult = findViewById<TextView>(R.id.textSearchResult)

        noSearchResult = findViewById(R.id.textNoSearchResult)
        rvSearch = findViewById(R.id.rvSearchList)
        progressSearch = findViewById(R.id.loadingSearch)
        setSupportActionBar(searchToolbar)
        setSearchText(searchResult, query)

        val retrofit = client
        apiService = retrofit!!.create(ApiService::class.java)

        val mPortraitLayoutManager = GridLayoutManager(this, 3)
        val mLandscapeLayoutManager = GridLayoutManager(this, 5)

        if(applicationContext.resources.configuration.orientation == Configuration.ORIENTATION_LANDSCAPE){
            rvSearch.layoutManager = mLandscapeLayoutManager
        } else {
            rvSearch.layoutManager = mPortraitLayoutManager
        }

        when (type) {
            "anime" -> {
                searchToolbar.title = "Search Anime"
                searchToolbar.subtitle = query

                searchAnimeAdapter = AnimeAdapter(searchAnimeResults, this)
                rvSearch.adapter = searchAnimeAdapter
                searchAnimeData(page)

                rvSearch.addOnScrollListener(object : RecyclerView.OnScrollListener() {
                    override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
                        super.onScrolled(recyclerView, dx, dy)
                        if (!recyclerView.canScrollVertically(1)) {
                            page++
                            searchAnimeData(page)
                        }
                    }
                })
            }
            "manga" -> {
                searchToolbar.title = "Search Manga"
                searchToolbar.subtitle = query

                searchMangaAdapter = MangaAdapter(searchMangaResults, this)
                rvSearch.adapter = searchMangaAdapter
                searchMangaData(page)

                rvSearch.addOnScrollListener(object : RecyclerView.OnScrollListener() {
                    override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
                        super.onScrolled(recyclerView, dx, dy)
                        if (!recyclerView.canScrollVertically(1)) {
                            page++
                            searchMangaData(page)
                        }
                    }
                })
            }
            "character" -> {
                searchToolbar.title = "Search Character"
                searchToolbar.subtitle = query

                searchCharacterAdapter = CharacterAdapter(searchCharacterResults, this)
                rvSearch.adapter = searchCharacterAdapter
                searchCharacterData(page)

                rvSearch.addOnScrollListener(object : RecyclerView.OnScrollListener() {
                    override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
                        super.onScrolled(recyclerView, dx, dy)
                        if (!recyclerView.canScrollVertically(1)) {
                            page++
                            searchCharacterData(page)
                        }
                    }
                })
            }
        }
    }

    private fun searchAnimeData(page: Int) {
        val limit = 15
        val call = apiService!!.searchAllAnime(page, limit, query)

        call!!.enqueue(object : Callback<AnimeResponse?> {
            override fun onResponse(
                call: Call<AnimeResponse?>,
                response: Response<AnimeResponse?>
            ) {
                if (response.body() != null) {
                    val oldCount = searchAnimeResults.size
                    progressSearch.visibility = View.GONE
                    rvSearch.visibility = View.VISIBLE
                    searchAnimeResults.addAll(response.body()!!.animeResults)
                    searchAnimeAdapter!!.notifyItemRangeInserted(oldCount, searchAnimeResults.size)
                } else if (searchAnimeResults.isEmpty()) {
                    progressSearch.visibility = View.GONE
                    noSearchResult.visibility = View.VISIBLE
                } else {
                    progressSearch.visibility = View.GONE
                }
            }

            override fun onFailure(call: Call<AnimeResponse?>, t: Throwable) {
                progressSearch.visibility = View.GONE
                Toast.makeText(
                    this@SearchActivity, "Fail to Fetch Data !!!",
                    Toast.LENGTH_SHORT
                ).show()
            }
        })
    }

    private fun searchMangaData(page: Int) {
        val limit = 15
        val call = apiService!!.searchAllManga(page, limit, query)

        call!!.enqueue(object : Callback<MangaResponse?> {
            override fun onResponse(
                call: Call<MangaResponse?>,
                response: Response<MangaResponse?>
            ) {
                if (response.body() != null) {
                    val oldCount = searchMangaResults.size
                    progressSearch.visibility = View.GONE
                    rvSearch.visibility = View.VISIBLE
                    searchMangaResults.addAll(response.body()!!.mangaResults)
                    searchMangaAdapter!!.notifyItemRangeInserted(oldCount, searchMangaResults.size)
                } else if (searchMangaResults.isEmpty()) {
                    progressSearch.visibility = View.GONE
                    noSearchResult.visibility = View.VISIBLE
                } else {
                    progressSearch.visibility = View.GONE
                }
            }

            override fun onFailure(call: Call<MangaResponse?>, t: Throwable) {
                progressSearch.visibility = View.GONE
                Toast.makeText(
                    this@SearchActivity, t.message + " cause : "
                            + t.cause, Toast.LENGTH_SHORT
                ).show()
            }
        })
    }

    private fun searchCharacterData(page: Int) {
        val limit = 15
        val call = apiService!!.searchAllCharacter(page, limit, query)

        call!!.enqueue(object : Callback<CharacterResponse?> {
            override fun onResponse(
                call: Call<CharacterResponse?>,
                response: Response<CharacterResponse?>
            ) {
                if (response.body() != null) {
                    val oldCount = searchCharacterResults.size
                    progressSearch.visibility = View.GONE
                    rvSearch.visibility = View.VISIBLE
                    searchCharacterResults.addAll(response.body()!!.characterResults)
                    searchCharacterAdapter!!.notifyItemRangeInserted(
                        oldCount,
                        searchCharacterResults.size
                    )
                } else if (searchCharacterResults.isEmpty()) {
                    progressSearch.visibility = View.GONE
                    noSearchResult.visibility = View.VISIBLE
                } else {
                    progressSearch.visibility = View.GONE
                }
            }

            override fun onFailure(call: Call<CharacterResponse?>, t: Throwable) {
                progressSearch.visibility = View.GONE
                Toast.makeText(
                    this@SearchActivity, "Fail to Fetch Data !!!",
                    Toast.LENGTH_SHORT
                ).show()
            }
        })
    }

    private fun setSearchText(tv: TextView, textValue: String?) {
        tv.text = HtmlCompat.fromHtml(
            "Search Result For : <b>' $textValue '</b>",
            HtmlCompat.FROM_HTML_MODE_LEGACY
        )
    }
}